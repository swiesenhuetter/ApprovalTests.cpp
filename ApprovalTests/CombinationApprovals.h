#pragma once

#include <type_traits>
#include "reporters/DefaultReporter.h"
#include "ApprovalTests/core/Reporter.h"
#include "ApprovalTests/utilities/CartesianProduct.h"
#include "Approvals.h"

namespace ApprovalTests
{
    namespace CombinationApprovals
    {
        namespace Detail
        {

            // Write out second or subsequent input value, with preceding comma and space
            struct print_input
            {
                std::ostream& out;
                template <class T> void operator()(const T& input)
                {
                    out << ", " << input;
                }
            };

            // Write out one row of output
            template <class Converter> struct serialize
            {
                std::ostream& out;
                Converter converter;
                template <class T, class... Ts>
                void operator()(T&& input1_, Ts&&... inputs)
                {
                    // First value is printed without trailing comma
                    out << "(" << input1_;
                    // Remaining values are printed with prefix of a comma
                    CartesianProduct::Detail::for_each(std::forward_as_tuple(inputs...),
                                                       print_input{out});
                    out << ") => " << converter(input1_, inputs...) << '\n';
                }
            };
        } // namespace Detail

        template <class Converter, class Container, class... Containers>
        void verifyAllCombinations(const Reporter& reporter,
                                   Converter&& converter,
                                   const Container& input0,
                                   const Containers&... inputs)
        {
            std::stringstream s;
            CartesianProduct::cartesian_product(
                Detail::serialize<Converter>{s, std::forward<Converter>(converter)},
                input0,
                inputs...);
            Approvals::verify(s.str(), reporter);
        }

        template <class Converter, class... Containers>
        ApprovalTests::Detail::EnableIfNotDerivedFromReporter<Converter>
        verifyAllCombinations(Converter&& converter, const Containers&... inputs)
        {
            verifyAllCombinations(
                DefaultReporter(), std::forward<Converter>(converter), inputs...);
        }

    } // namespace CombinationApprovals
} // namespace ApprovalTests
