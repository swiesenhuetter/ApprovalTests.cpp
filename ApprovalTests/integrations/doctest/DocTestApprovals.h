#pragma once

#include "ApprovalTests/namers/ApprovalTestNamer.h"
#include "ApprovalTests/integrations/CheckFileMacroIsAbsolute.h"

#if defined(APPROVALS_DOCTEST_EXISTING_MAIN)
#define APPROVALS_DOCTEST
#define DOCTEST_CONFIG_IMPLEMENT
#elif defined(APPROVALS_DOCTEST)
#define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
#endif

#ifdef APPROVALS_DOCTEST

// begin-snippet: required_header_for_doctest
#include <doctest/doctest.h>
// end-snippet

namespace ApprovalTests
{
    // anonymous namespace to prevent compiler -Wsubobject-linkage warnings
    // This is OK as this code is only compiled on main()
    namespace
    {
        struct AbstractReporter : doctest::IReporter
        {
            virtual void report_query(const doctest::QueryData&) override
            {
            }
            // called when the whole test run starts
            virtual void test_run_start() override
            {
            }

            // called when the whole test run ends (caching a pointer to the input doesn't make sense here)
            virtual void test_run_end(const doctest::TestRunStats&) override
            {
            }

            // called when a test case is started (safe to cache a pointer to the input)
            virtual void test_case_start(const doctest::TestCaseData&) override
            {
            }

#if 20305 <= DOCTEST_VERSION
            // called when a test case is reentered because of unfinished subcases (safe to cache a pointer to the input)
            virtual void test_case_reenter(const doctest::TestCaseData&) override
            {
            }
#endif

            // called when a test case has ended
            virtual void test_case_end(const doctest::CurrentTestCaseStats&) override
            {
            }

            // called when an exception is thrown from the test case (or it crashes)
            virtual void test_case_exception(const doctest::TestCaseException&) override
            {
            }

            // called whenever a subcase is entered (don't cache pointers to the input)
            virtual void subcase_start(const doctest::SubcaseSignature&) override
            {
            }

            // called whenever a subcase is exited (don't cache pointers to the input)
            virtual void subcase_end() override
            {
            }

            // called for each assert (don't cache pointers to the input)
            virtual void log_assert(const doctest::AssertData&) override
            {
            }

            // called for each message (don't cache pointers to the input)
            virtual void log_message(const doctest::MessageData&) override
            {
            }

            // called when a test case is skipped either because it doesn't pass the filters, has a skip decorator
            // or isn't in the execution range (between first and last) (safe to cache a pointer to the input)
            virtual void test_case_skipped(const doctest::TestCaseData&) override
            {
            }
        };

        struct DocTestApprovalListener : AbstractReporter
        {
            TestName currentTest;

            // constructor has to accept the ContextOptions by ref as a single argument
            explicit DocTestApprovalListener(const doctest::ContextOptions& /*in*/)
            {
            }

            void test_case_start(const doctest::TestCaseData& testInfo) override
            {

                currentTest.sections.emplace_back(testInfo.m_name);
                currentTest.setFileName(testInfo.m_file);
                ApprovalTestNamer::currentTest(&currentTest);
            }

            void test_case_end(const doctest::CurrentTestCaseStats& /*in*/) override
            {

                while (!currentTest.sections.empty())
                {
                    currentTest.sections.pop_back();
                }
            }

            void subcase_start(const doctest::SubcaseSignature& signature) override
            {

#if DOCTEST_VERSION >= 20307
                currentTest.sections.emplace_back(signature.m_name.c_str());
#else
                currentTest.sections.emplace_back(signature.m_name);
#endif
            }

            void subcase_end() override
            {

                currentTest.sections.pop_back();
            }
        };
    }
}

REGISTER_LISTENER("approvals", 0, ApprovalTests::DocTestApprovalListener);

#endif // APPROVALS_DOCTEST
