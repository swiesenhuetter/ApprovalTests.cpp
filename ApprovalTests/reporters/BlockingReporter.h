#pragma once

#include "ApprovalTests/core/Reporter.h"
#include "../utilities/MachineBlocker.h"

#include <memory>
#include <utility>

namespace ApprovalTests
{
    class BlockingReporter : public Reporter
    {
    private:
        std::shared_ptr<Blocker> blocker;

        BlockingReporter() = delete;

    public:
        explicit BlockingReporter(std::shared_ptr<Blocker> blocker_)
            : blocker(std::move(blocker_))
        {
        }

        static std::shared_ptr<BlockingReporter>
        onMachineNamed(const std::string& machineName)
        {
            auto machineBlocker = std::make_shared<MachineBlocker>(
                MachineBlocker::onMachineNamed(machineName));
            return std::make_shared<BlockingReporter>(machineBlocker);
        }

        static std::shared_ptr<BlockingReporter>
        onMachinesNotNamed(const std::string& machineName)
        {
            auto machineBlocker = std::make_shared<MachineBlocker>(
                MachineBlocker::onMachinesNotNamed(machineName));
            return std::make_shared<BlockingReporter>(machineBlocker);
        }

        virtual bool report(std::string /*received*/,
                            std::string /*approved*/) const override
        {
            return blocker->isBlockingOnThisMachine();
        }
    };
}
