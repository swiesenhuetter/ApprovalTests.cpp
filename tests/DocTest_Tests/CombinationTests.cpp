#include "doctest/doctest.h"
#include <map>
#include <vector>
#include <string>
#include <ApprovalTests/reporters/BlockingReporter.h>
#include "ApprovalTests/CombinationApprovals.h"
#include "reporters/FakeReporter.h"

using namespace ApprovalTests;

TEST_CASE("YouCanVerifyCombinationsOf1")
{
    std::vector<std::string> words{"hello", "world"};
    CombinationApprovals::verifyAllCombinations(
        [](const std::string& s) { return s + "!"; }, words);
}

TEST_CASE("YouCanVerifyCombinationsOf1WithTemplateParameters")
{
    std::vector<std::string> words{"hello", "world"};
    CombinationApprovals::verifyAllCombinations(
        [](const std::string& s) { return s + "!"; }, words);
}

FrontLoadedReporterDisposer clearFrontLoadedReporter()
{
    return Approvals::useAsFrontLoadedReporter(
        BlockingReporter::onMachineNamed("safadfasdfas"));
}

TEST_CASE("YouCanVerifyCombinationsOf1Reports")
{
    auto d = clearFrontLoadedReporter();
    std::vector<std::string> words{"hello", "world"};
    FakeReporter reporter;
    try
    {
        CombinationApprovals::verifyAllCombinations(
            reporter, [](const std::string& s) { return s + "!"; }, words);
    }
    catch (const ApprovalException&)
    {
        // ignore
    }
    REQUIRE(reporter.called == true);
}

// begin-snippet: YouCanVerifyCombinationsOf2
TEST_CASE("YouCanVerifyCombinationsOf2")
{
    std::vector<std::string> v{"hello", "world"};
    std::vector<int> numbers{1, 2, 3};
    CombinationApprovals::verifyAllCombinations(
        [](std::string s, int i) {
            return std::string("(") + s + ", " + std::to_string(i) + ")";
        },
        v,
        numbers);
}
// end-snippet

// begin-snippet: CombinationsStartingPoint
TEST_CASE("CombinationsStartingPoint")
{
    std::vector<std::string> inputs1{"input1.value1", "input1.value2"};
    std::vector<std::string> inputs2{"input2.value1", "input2.value2", "input2.value3"};
    CombinationApprovals::verifyAllCombinations(
        [](auto /*input1*/, auto /*input2*/) { return "placeholder"; }, inputs1, inputs2);
}
// end-snippet

TEST_CASE("YouCanVerifyCombinationsOf9")
{
    std::vector<std::string> letters{"a", "b"};
    CombinationApprovals::verifyAllCombinations(
        [](const std::string& s1,
           const std::string& s2,
           const std::string& s3,
           const std::string& s4,
           const std::string& s5,
           const std::string& s6,
           const std::string& s7,
           const std::string& s8,
           const std::string& s9) { return s1 + s2 + s3 + s4 + s5 + s6 + s7 + s8 + s9; },
        letters,
        letters,
        letters,
        letters,
        letters,
        letters,
        letters,
        letters,
        letters);
}

TEST_CASE("YouCanVerifyCombinationsOf9Reports")
{
    auto d = clearFrontLoadedReporter();
    FakeReporter reporter;
    try
    {
        std::vector<std::string> letters{"a", "b"};
        CombinationApprovals::verifyAllCombinations(
            reporter,
            [](const std::string& s1,
               const std::string& s2,
               const std::string& s3,
               const std::string& s4,
               const std::string& s5,
               const std::string& s6,
               const std::string& s7,
               const std::string& s8,
               const std::string& s9) {
                return s1 + s2 + s3 + s4 + s5 + s6 + s7 + s8 + s9;
            },
            letters,
            letters,
            letters,
            letters,
            letters,
            letters,
            letters,
            letters,
            letters);
    }
    catch (const ApprovalException&)
    {
        // ignore
    }
    REQUIRE(reporter.called == true);
}

TEST_CASE("YouCanVerifyCombinationsOf10")
{
    std::vector<std::string> letters{"a", "b"};
    CombinationApprovals::verifyAllCombinations(
        [](const std::string& s1,
           const std::string& s2,
           const std::string& s3,
           const std::string& s4,
           const std::string& s5,
           const std::string& s6,
           const std::string& s7,
           const std::string& s8,
           const std::string& s9,
           const std::string& s10) {
            return s1 + s2 + s3 + s4 + s5 + s6 + s7 + s8 + s9 + s10;
        },
        letters,
        letters,
        letters,
        letters,
        letters,
        letters,
        letters,
        letters,
        letters,
        letters);
}
