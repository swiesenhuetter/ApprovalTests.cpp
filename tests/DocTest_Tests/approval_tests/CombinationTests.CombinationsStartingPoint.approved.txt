(input1.value1, input2.value1) => placeholder
(input1.value1, input2.value2) => placeholder
(input1.value1, input2.value3) => placeholder
(input1.value2, input2.value1) => placeholder
(input1.value2, input2.value2) => placeholder
(input1.value2, input2.value3) => placeholder

